Signal/src/ViewControllers/DebugUI/DebugUINotifications.swift

```diff
@@ -0,0 +1,117 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import SignalServiceKit
+import SignalMessaging
+
+class DebugUINotifications: DebugUIPage {
+
+    // MARK: Dependencies
+
+    var notificationsManager: NotificationsManager {
+        return SignalApp.shared().notificationsManager
+    }
+    var notificationsAdapter: CallNotificationsAdapter {
+        return SignalApp.shared().callService.notificationsAdapter
+    }
+    var messageSender: MessageSender {
+        return Environment.current().messageSender
+    }
+    var contactsManager: OWSContactsManager {
+        return Environment.current().contactsManager
+    }
+
+    // MARK: Overrides
+
+    override func name() -> String {
+        return "Notifications"
+    }
+
+    override func section(thread aThread: TSThread?) -> OWSTableSection? {
+        guard let thread = aThread else {
+            owsFail("\(logTag) Notifications must specify thread.")
+            return nil
+        }
+
+        var sectionItems = [
+            OWSTableItem(title:"Last Incoming Message") {
+                Logger.info("\(self.logTag) scheduling notification for incoming message.")
```

Comment:
These `self` references are going to create retain cycles that will leak this UI.  I don't care about the debug UI, but we don't want to use this idiom in production code.  I wonder if there's a convenient way to do something safe in production code?

Reply:
Thanks, fixed!